File: agent/src/nexus/server/routes/vault_history.py
"""Routes for vault history (opt-in, git-backed): /vault/history*."""

from __future__ import annotations

import asyncio

from fastapi import APIRouter, HTTPException, status

router = APIRouter()


@router.get("/vault/history/status")
async def vault_history_status() -> dict:
    from ... import vault_history
    return await asyncio.to_thread(vault_history.status)


@router.post("/vault/history/enable")
async def vault_history_enable() -> dict:
    from ... import vault_history
    try:
        return await asyncio.to_thread(vault_history.enable)
    except vault_history.HistoryError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.post("/vault/history/disable")
async def vault_history_disable() -> dict:
    from ... import vault_history
    return await asyncio.to_thread(vault_history.disable)


@router.get("/vault/history")
async def vault_history_log(path: str | None = None, limit: int = 100) -> dict:
    from ... import vault_history
    if limit < 1 or limit > 1000:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="limit must be 1..1000")
    try:
        commits = await asyncio.to_thread(vault_history.log, path=path, limit=limit)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc))
    return {
        "path": path,
        "commits": [
            {
                "sha": c.sha,
                "timestamp": c.timestamp,
                "message": c.message,
                "action": c.action,
            }
            for c in commits
        ],
    }


@router.post("/vault/history/undo")
async def vault_history_undo(body: dict) -> dict:
    from ... import vault_history
    path = body.get("path")
    if not isinstance(path, str) or not path:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="`path` is required")
    try:
        result = await asyncio.to_thread(vault_history.undo, path)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc))
    return {
        "undone": result.undone,
        "reason": result.reason,
        "commit": result.commit,
        "restored_from": result.restored_from,
        "paths": result.paths or [],
    }


@router.post("/vault/history/purge")
async def vault_history_purge(body: dict | None = None) -> dict:
    from ... import vault_history
    before_iso = (body or {}).get("before_iso") if body else None
    return await asyncio.to_thread(vault_history.purge, before_iso=before_iso)

File: ui/src/api/vaultHistory.ts
// API client for the opt-in vault history (git-backed) — see vault_history.py.
import { BASE } from "./base";

export type VaultHistoryAction = "write" | "delete" | "move" | "undo" | "enable" | "other";

export interface VaultHistoryCommit {
  sha: string;
  /** Unix seconds. */
  timestamp: number;
  message: string;
  action: VaultHistoryAction;
}

export interface VaultHistoryStatus {
  enabled: boolean;
  repo_exists: boolean;
  git_available: boolean;
  commit_count: number;
  last_commit: VaultHistoryCommit | null;
}

export interface VaultUndoResult {
  undone: boolean;
  /** Set when ``undone`` is false: e.g. "no_history", "disabled", "no_repo". */
  reason: string | null;
  commit: string | null;
  restored_from: string | null;
  paths: string[];
}

export async function getVaultHistoryStatus(): Promise<VaultHistoryStatus> {
  const res = await fetch(`${BASE}/vault/history/status`);
  if (!res.ok) throw new Error(`Vault history status error: ${res.status}`);
  return res.json() as Promise<VaultHistoryStatus>;
}

export async function enableVaultHistory(): Promise<VaultHistoryStatus> {
  const res = await fetch(`${BASE}/vault/history/enable`, { method: "POST" });
  if (!res.ok) {
    const detail = await res.text().catch(() => "");
    throw new Error(`Vault history enable error: ${res.status} ${detail}`);
  }
  return res.json() as Promise<VaultHistoryStatus>;
}

export async function disableVaultHistory(): Promise<VaultHistoryStatus> {
  const res = await fetch(`${BASE}/vault/history/disable`, { method: "POST" });
  if (!res.ok) throw new Error(`Vault history disable error: ${res.status}`);
  return res.json() as Promise<VaultHistoryStatus>;
}

export async function getVaultHistory(
  path?: string,
  limit = 100,
): Promise<VaultHistoryCommit[]> {
  const qs = new URLSearchParams();
  if (path) qs.set("path", path);
  qs.set("limit", String(limit));
  const res = await fetch(`${BASE}/vault/history?${qs.toString()}`);
  if (!res.ok) throw new Error(`Vault history log error: ${res.status}`);
  const data = await res.json() as { commits: VaultHistoryCommit[] };
  return data.commits;
}

export async function undoVaultPath(path: string): Promise<VaultUndoResult> {
  const res = await fetch(`${BASE}/vault/history/undo`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ path }),
  });
  if (!res.ok) throw new Error(`Vault undo error: ${res.status}`);
  return res.json() as Promise<VaultUndoResult>;
}

export async function purgeVaultHistory(): Promise<{ ok: boolean; reason?: string }> {
  const res = await fetch(`${BASE}/vault/history/purge`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({}),
  });
  if (!res.ok) throw new Error(`Vault history purge error: ${res.status}`);
  return res.json() as Promise<{ ok: boolean; reason?: string }>;
}
